### ISSUE TITLE
regression 3.0.5: tsql `RF03` bug. `INSERT` must not contain table reference, `$ACTION` must not contain table reference

### ISSUE THREAD

**Issue by fmms** [sqlfluff/sqlfluff#5980]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

I tried to format a tsql Statement, It was broken afterwards.

### Expected Behaviour

Statement still executes afterwards

### Observed Behaviour

`MRG` alias is introduced in places where it is not valid.

### How to reproduce

take the following statement

```sql
INSERT INTO [DIL].[md_GLAccount_HIST] (
  [ChartOfAccounts_KEY]
  , [GLAccount_KEY]
  , [ValidFrom]
  , [ValidTo]
  , [GLAccount_DESC_short]
)
SELECT
  MRG.[ChartOfAccounts_KEY]
  , MRG.[GLAccount_KEY]
  , MRG.[ValidFrom]
  , MRG.[ValidTo]
  , MRG.[GLAccount_DESC_short]
FROM (
  MERGE [DIL].[md_GLAccount_HIST] WITH (TABLOCK) t
  USING [INTERNAL_MERGE].[DIL_md_GLAccount_HIST] s
    ON (
      t.[ChartOfAccounts_KEY] = s.[ChartOfAccounts_KEY]
      AND t.[GLAccount_KEY] = s.[GLAccount_KEY]
    )
  WHEN NOT MATCHED BY TARGET
    THEN
    INSERT (
      [ChartOfAccounts_KEY]
      , [GLAccount_KEY]
      , [ValidFrom]
      , [ValidTo]
      , [GLAccount_DESC_short]
    )
    VALUES (
      s.[ChartOfAccounts_KEY]
      , s.[GLAccount_KEY]
      , '19900101'
      , '99991231'
      , s.[GLAccount_DESC_short]
    )
  OUTPUT
    $ACTION AS [Action_OUT]
    , s.[ChartOfAccounts_KEY]
    , s.[GLAccount_KEY]
    , '19900101'
    , '99991231'
    , s.[GLAccount_DESC_short]
) MRG
WHERE
  MRG.Action_OUT = 'UPDATE'
  AND MRG.[ChartOfAccounts_KEY] IS NOT NULL
  AND MRG.[GLAccount_KEY] IS NOT NULL;
```

Format the above statement:
```ps1
PS C:\Users\fmoeller\Desktop\RepositoryVorlage>  sqlfluff fix --dialect tsql  .\query.sql  --fixed-suffix fixed
==== finding fixable violations ====
== [query.sql] FAIL
L:  24 | P:   7 | RF03 | Unqualified reference '[ChartOfAccounts_KEY]' found in
                       | single table select. [references.consistent]
L:  25 | P:   9 | RF03 | Unqualified reference '[GLAccount_KEY]' found in single
                       | table select. [references.consistent]
L:  26 | P:   9 | RF03 | Unqualified reference '[ValidFrom]' found in single
                       | table select. [references.consistent]
L:  27 | P:   9 | RF03 | Unqualified reference '[ValidTo]' found in single table
                       | select. [references.consistent]
L:  28 | P:   9 | RF03 | Unqualified reference '[GLAccount_DESC_short]' found in
                       | single table select. [references.consistent]
L:  38 | P:   5 | RF03 | Unqualified reference '$ACTION' found in single table
                       | select. [references.consistent]
== [query.sql] FIXED
6 fixable linting violations found
  [13 unfixable linting violations found]
```

The result is the following code:
```sql
INSERT INTO [DIL].[md_GLAccount_HIST] (
  [ChartOfAccounts_KEY]
  , [GLAccount_KEY]
  , [ValidFrom]
  , [ValidTo]
  , [GLAccount_DESC_short]
)
SELECT
  MRG.[ChartOfAccounts_KEY]
  , MRG.[GLAccount_KEY]
  , MRG.[ValidFrom]
  , MRG.[ValidTo]
  , MRG.[GLAccount_DESC_short]
FROM (
  MERGE [DIL].[md_GLAccount_HIST] WITH (TABLOCK) t
  USING [INTERNAL_MERGE].[DIL_md_GLAccount_HIST] s
    ON (
      t.[ChartOfAccounts_KEY] = s.[ChartOfAccounts_KEY]
      AND t.[GLAccount_KEY] = s.[GLAccount_KEY]
    )
  WHEN NOT MATCHED BY TARGET
    THEN
    INSERT (
      MRG.[ChartOfAccounts_KEY]
      , MRG.[GLAccount_KEY]
      , MRG.[ValidFrom]
      , MRG.[ValidTo]
      , MRG.[GLAccount_DESC_short]
    )
    VALUES (
      s.[ChartOfAccounts_KEY]
      , s.[GLAccount_KEY]
      , '19900101'
      , '99991231'
      , s.[GLAccount_DESC_short]
    )
  OUTPUT
    MRG.$ACTION AS [Action_OUT]
    , s.[ChartOfAccounts_KEY]
    , s.[GLAccount_KEY]
    , '19900101'
    , '99991231'
    , s.[GLAccount_DESC_short]
) MRG
WHERE
  MRG.Action_OUT = 'UPDATE'
  AND MRG.[ChartOfAccounts_KEY] IS NOT NULL
  AND MRG.[GLAccount_KEY] IS NOT NULL;
```
that does not execute anymore.

### Dialect

tsql

### Version

```
==== sqlfluff ====
sqlfluff:                3.0.7 python:                 3.11.9
implementation:        cpython verbosity:                   1
dialect:                  tsql templater:                 raw
rules:                                all
```

### Configuration

none

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by fmms** [comment]
Formatting the above code with sqlfluff 3.0.4 does not cause the damage as seen with 3.0.5.

@keraion , @alanmcruickshank might this be introduced by #5726 ?

**Comment by keraion** [comment]
I see where the issue is. I'll pull something together when I have a chance later.
